### ISSUE TITLE
ZipMap: ([x], x → y) → [(x, y)]

### ISSUE THREAD

**Issue by atifaziz** [morelinq/MoreLINQ#661]
This was [initially proposed](https://github.com/morelinq/MoreLINQ/issues/624#issuecomment-549077029) by @Orace as part of the discussion in #624 but I am adding it here with the proposed name of `ZipMap`:

```c#
public static IEnumerable<(TSource Item, TResult Result)>
    ZipMap<TSource, TResult>(this IEnumerable<TSource> source, Func<TSource, TResult> func)
```

Element names are up for discussion.

**Comment by declard** [comment]
Can't this be a composition of functions like `IEnumerable<(A Left, A Right)> SelectPairs(IEnumerable<A> source)` and `IEnumerable<(A Left, C RIght)> MapRight<A, B, C>(IEnumerable<(A Left, B Right)> source, Func<B, C> f)` which may become a good generic addition to MoreLinq?

**Comment by atifaziz** [comment]
@declard Interesting idea! I like the generalization.

**Comment by viceroypenguin** [comment]
Interesting idea, but I'm not sure it adds value. Is there a use-case for `SelectPairs()` outside of `ZipMap`; that is, is there a reason someone would want a single value as both parts of a tuple other than for use with a `MapRight` method? 

Similarly, while I can see potential use-cases, I'm not sure that `MapRight()` (or an expected matching `MapLeft()`) is something that would be used often. Even when used in queries, `ValueTuples` are generally used wholly, not partially, in my experience. As such, it looks like a solution in search of a problem.

`ZipMap` is a fairly simple function in the first place, basically `.Select(x => (x, func(x)))`, but eliminating the closure. I'm not sure that further generalization would be helpful.

**Comment by declard** [comment]
Agree, `ValueTuple<,>` is not the best candidate. We have SelectPairs, MapKeys, and MapValues working with `KVP<,>` on a project, which allow things like
```csharp
collection
    .SelectPairs().MapValues(SomeFuncReturningOptionals).CollectSomeOptValues().MapValues(AnotherFunc).ToDictionary()`, very convenient.

If closure is in concern, then why not introducing something generic like
```csharp
IEnumerable<TResult> Select<TElement, TPayload, TResult>(
    this IEnumerable<TElement> collection,
    TPayload payload,
    Func<TElement, TPayload, TResult> func)
{
    using var iterator = collection.GetEnumerator();

    while (iterator.MoveNext())
    {
        yield return func(iterator.Current, payload);
    }
}
```
?

`collection.Select(func, (x, func) => (x, func(x)))` doesn't look as nice as `collection.ZipMap(func)`, unless the `(x, func) => (x, func(x))` part is abstracted into a function on its own, but it's still strictly more powerful and has exactly same memory footprint

**Comment by viceroypenguin** [comment]
> > As such, it looks like a solution in search of a problem.
> 
> What would that problem be?

I don't know; that's why I made that statement. The solution given by `MapRight()` is to execute a projection on part of a tuple while keeping the other element of the tuple as-is. My statement is that I don't know what problem this solves, so it is a solution in search of a problem.

However, @declard has offered some code where such a method may be useful. If one is keeping a source value as well as the net transformation of that source, then such a method would allow a simpler projection.

@declard My concern was never with the closure, `ValueTuple`, or with `ZipMap`. In fact, I have `ZipMap` implemented in elsewhere - it is definitely cleaner to have `source.ZipMap(func)` than `source.Select(x => (x, func(x)))`. My concern was with the additional level of abstraction suggested via `SelectPair()` and `MapRight()`. I would still argue, even in your example, that `SelectPair` specifically is not an inherently useful method. It would be better to use `ZipMap` for the first projection, and then `MapRight` for any future projections. 

I would still want to argue the cost-benefit of a method like `MapRight`, since it is a proxy for `source.Select(x => (x.Left, Right: func(x.Right)))`; though if `ZipMap` passes the bar for usefulness (which it does in my book), then I would be hard-pressed to counter-argue that `MapRight` does not pass the bar.

**Comment by declard** [comment]
By the way, `collection.ZipMap(func)` on KVPs is expressed as `collection.SelectPairs().MapValues(func)`

**Comment by declard** [comment]
> Or is it only a useful method because of `MapValues()`?

well, because of `MapKeys` too, maybe some other things which take KVPs, like `ToDictionary`

**Comment by atifaziz** [comment]
> I have `ZipMap` implemented in elsewhere - it is definitely cleaner to have `source.ZipMap(func)` than `source.Select(x => (x, func(x)))`. My concern was with the additional level of abstraction suggested via `SelectPair()` and `MapRight()`. I would still argue, even in your example, that `SelectPair` specifically is not an inherently useful method. It would be better to use `ZipMap` for the first projection, and then `MapRight` for any future projections.
> 
> I would still want to argue the cost-benefit of a method like `MapRight`, since it is a proxy for `source.Select(x => (x.Left, Right: func(x.Right)))`; though if `ZipMap` passes the bar for usefulness (which it does in my book), then I would be hard-pressed to counter-argue that `MapRight` does not pass the bar.

Thanks for expanding. While I said [I like the generalizations](https://github.com/morelinq/MoreLINQ/issues/661#issuecomment-1487635677), I don't think `SelectPairs` and `MapRight` have a place in MoreLINQ. In fact, I could extend that to `Match` (#624) and `ZipMap` here too, which is why I haven't jumped at adding those. These are all very trivial implementations that seem to have to do more with lifting of functions into wrapped or inner structures while maintaining the wrapper or outer structure (the sequence) than something algorithmically interesting for sequences. I'm not debating their utility for some people, but just that they probably belong in a separately maintained library, may be one centered around tuples (like I did for `KeyValuePair<,>` and sequences thereof with [**KeyValuePairs**](https://github.com/atifaziz/KeyValuePairs)). `MapRight` is especially dangerous because it could lead to scope creep (apply a function to an element of N-tuple or apply the N-tuple as arguments to a function, etc.). I had imagined that `Match` and `ZipMap` could be interesting for quick and interactive explorations in tools like [**LINQPad**](https://www.linqpad.net/)), but I use the query syntax whenever possible and almost always create tuples with meaningful element names for readability (at which point you're back to relying on `select`/`Select`), so I've not needed or used them personally, but then again, this is not about just me. I'm happy to close these (considering they've been open for comment for a while) unless someone can demonstrate a strong case, especially for `ZipMap` that's the topic of this issue.
